Remind user to record stream on vehicle arming

ArturoManzoli opened the pull request:
- On first time arming the vehicle there will be a popup asking the user to configure the auto recording parameters;
- Later, the user can find this settings at the Video Configurations menu;
- Persistent and non persistent settings can be applied. If `remember my choice` checkbox is marked, settings will persist.

https://github.com/user-attachments/assets/4e525f14-5a75-4925-81c1-725ae34c6c54

Closes #947

rafaellehmkuhl commented:
Super cool!

ArturoManzoli commented:
Last patch:
- Added an option to set a reminder for starting stream recordings. The reminder can have a custom delay relative to vehicle arming;
- The reminder has a distinct visual style compared to regular snackbar messages.

![image](https://github.com/user-attachments/assets/b93cdcd9-0ad4-4ab6-937c-fcf172c90588)

https://github.com/user-attachments/assets/005a0e2a-1e71-40c9-9bd8-ccfe4ce51f7d

ES-Alexander commented:
Will review tomorrow, but just flagging that this shouldn’t occur if there are no video streams available - I haven’t checked whether that case is already handled.

ES-Alexander reviewed (approved):
Just got around to this - sorry!

From some (admittedly limited) testing the reminder seems to work as expected, including not showing up if there's no video stream connected 👍 

I didn't end up seeing a configuration popup on the first arm after a video stream was available, but that's likely because I had already gone in and played with the settings while I didn't have a video stream.

rafaellehmkuhl commented on src/stores/video.ts:
```diff
@@ -932,6 +942,46 @@ export const useVideoStore = defineStore('video', () => {
+    },
+    { immediate: true }
+  )
+  console.log('🚀 ~ currentAutoRecordOptions.value:', currentAutoRecordOptions.value)
```
To remove.

rafaellehmkuhl reviewed (changes requested):
I don't know why, but the reminder is not showing here:


https://github.com/user-attachments/assets/9c92ce19-d733-4d56-aa94-1b5d74ff39f1

PS: I did check if I was running the right version by going to the video config page and finding the new dialog button there.

rafaellehmkuhl commented on src/stores/video.ts:
```diff
@@ -57,7 +58,16 @@ export const useVideoStore = defineStore('video', () => {
   const autoProcessVideos = useBlueOsStorage('cockpit-auto-process-videos', true)
   const lastRenamedStreamName = ref('')
-
+  const cockpitAutoRecordStreams = useBlueOsStorage<AutoRecordVideoStreams>('cockpit-auto-record-streams', {
```
This key should be on a variable somewhere (probably the types file) and exported so the other components using it can import and prevent mismatches.

ArturoManzoli replied:
fixed!

rafaellehmkuhl commented on src/stores/video.ts:
```diff
@@ -932,6 +942,46 @@ export const useVideoStore = defineStore('video', () => {
+      if (isArmed && streamsCorrespondency.value.length > 0) {
+        if (currentAutoRecordOptions.value.autoRecordOption === 'all') {
+          startRecordingAllStreams()
+          openSnackbar({
```
This one and the other below can be on single lines.

ArturoManzoli replied:
Fixed

rafaellehmkuhl replied:
Are you sure? It looks like the same as before here:

![image](https://github.com/user-attachments/assets/f8da1db3-992b-43e6-afdf-13f599ebf13b)

rafaellehmkuhl reviewed (changes requested):
If a user opens Cockpit and the vehicle is already armed, is the reminder going to kick in? The same for the auto-recording?

Since the feature isn't working here, I couldn't check this case, but I suspect it could be not doing it based on the `watch` logic, but I can be wrong.

rafaellehmkuhl commented:
> I didn't end up seeing a configuration popup on the first arm after a video stream was available, but that's likely because I had already gone in and played with the settings while I didn't have a video stream.

I think there's actually a bug around the first usage. It didn't work here as well, and I didn't have accessed the configuration page yet.

@ArturoManzoli it worked after I've accessed the reminder config dialog for the first time. I think it's something related to that.

rafaellehmkuhl commented:
> If a user opens Cockpit and the vehicle is already armed, is the reminder going to kick in? The same for the auto-recording?
> 
> Since the feature isn't working here, I couldn't check this case, but I suspect it could be not doing it based on the `watch` logic, but I can be wrong.

After opening the dialog config and making the feature work, I could confirm that it does remind if Cockpit is open with the vehicle already online! Great!

ArturoManzoli commented:
> > I didn't end up seeing a configuration popup on the first arm after a video stream was available, but that's likely because I had already gone in and played with the settings while I didn't have a video stream.
> 
> I think there's actually a bug around the cold usage. It didn't work here as well, and I didn't have accessed the configuration page yet.
> 
> @ArturoManzoli it worked after I've accessed the reminder config dialog for the first time. I think it's something related to that.

There was indeed a problem. It's fixed now!

rafaellehmkuhl reviewed (changes requested):
Almost everything is working. The only thing that is not is that if I open Cockpit with the vehicle already armed, it does not auto-record, nor sends recording reminders.

ArturoManzoli commented:
> Almost everything is working. The only thing that is not is that if I open Cockpit with the vehicle already armed, it does not auto-record, nor sends recording reminders.

I misunderstood your first message about this issue.
Now it's implemented correctly :)

rafaellehmkuhl commented:
> > Almost everything is working. The only thing that is not is that if I open Cockpit with the vehicle already armed, it does not auto-record, nor sends recording reminders.
> 
> I misunderstood your first message about this issue. Now it's implemented correctly :)


https://github.com/user-attachments/assets/15bd3aac-793c-4ea4-b029-1ed250ec279a

Problem appears to still be there.

rafaellehmkuhl reviewed (changes requested):
> > Almost everything is working. The only thing that is not is that if I open Cockpit with the vehicle already armed, it does not auto-record, nor sends recording reminders.
> 
> I misunderstood your first message about this issue. Now it's implemented correctly :)


https://github.com/user-attachments/assets/15bd3aac-793c-4ea4-b029-1ed250ec279a

Problem appears to still be there.

ArturoManzoli commented:
> > > Almost everything is working. The only thing that is not is that if I open Cockpit with the vehicle already armed, it does not auto-record, nor sends recording reminders.
> > 
> > 
> > I misunderstood your first message about this issue. Now it's implemented correctly :)
> 
>  Kapture.2025-03-27.at.12.58.14.mp4 
> Problem appears to still be there.

Ah, that's because the stream wasn't available yet when Cockpit finished booting.
I'll take a look

rafaellehmkuhl commented on src/assets/defaults.ts:
```diff
@@ -901,3 +901,11 @@ export const defaultSensorDataloggerProfile: OverlayGrid = {
 }
+
+export const defaultAutoRecordingOptions = {
+  autoRecordOption: undefined,
```
`autoRecordMode` seems like a better choice of name.

Whenever you have an object with a name and a property with the same name or very similar (e.g.: AutoRecordOptions > autoRecordOption), it usually means there's an architectural problem (not the case here) or a naming problem.

rafaellehmkuhl replied:
I think you got it wrong. The object should stay as `autoRecordOptions` (as there are several options inside the object) and the property is the one to be renamed to `autoRecordMode` (as it has modes, like `all`, `none` and `selected`).

rafaellehmkuhl commented on src/stores/video.ts:
```diff
@@ -57,7 +59,14 @@ export const useVideoStore = defineStore('video', () => {
   const autoProcessVideos = useBlueOsStorage('cockpit-auto-process-videos', true)
   const lastRenamedStreamName = ref('')
-
+  const cockpitAutoRecordStreams = useBlueOsStorage<AutoRecordVideoStreams>(
+    'cockpit-auto-record-streams',
+    defaultAutoRecordingOptions
+  )
+  const currentAutoRecordOptions = ref(
+    cockpitAutoRecordStreams.value || { autoRecordOption: undefined, selectedStreams: [] }
+  )
```
Maybe this was a leftover from a first implementation try? Both those objects seem like doing the same thing. The value of one is actually used as the default for the other.

rafaellehmkuhl replied:
Got it. Makes more sense now. In this case they would be better named something like `momentaryAutoRecordOptions` and `persistentAutoRecordOptions`, so it's clear they are the same but for different situations.

One question: does it make sense to even have this "remember my choice" option? Shouldn't it be like that by default, and if the user decides to disable it at some point, they do it explicitly? Not something to hold the PR thought.

ArturoManzoli replied:
It might not be used as often as the 'remember my choice' option, but when following the suggestion from Willian Galvani's original issue (which included the `remember my choice` option), I thought it might be useful to on certain situations. 
I saw as being quite annoying for some users to always record the streams on vehicle arming, on every situation.  For that, the option to `do this only once` was designed.

rafaellehmkuhl replied:
@ArturoManzoli the comment about the naming is still relevant nonetheless.

rafaellehmkuhl reviewed (changes requested):
Everything is working fine now.

There are some things that can be improved in the code listed below.

rafaellehmkuhl commented on src/stores/video.ts:
```diff
@@ -932,6 +941,85 @@ export const useVideoStore = defineStore('video', () => {
+    (isArmed, oldVal) => {
+      if (oldVal !== undefined && oldVal && !isArmed && currentAutoRecordOptions.value.stopRecordingOnDisarm) {
+        stopRecordingAllStreams()
+        openSnackbar({
```
This snackbar can be a one liner.

rafaellehmkuhl commented on src/stores/video.ts:
```diff
@@ -932,6 +941,85 @@ export const useVideoStore = defineStore('video', () => {
+          const streamEntry = streamsCorrespondency.value.find((stream) => stream.name === name)
+          if (streamEntry) {
+            startRecording(streamEntry.externalId)
+            openSnackbar({
```
This can also be a one liner.

rafaellehmkuhl commented on src/stores/video.ts:
```diff
@@ -932,6 +941,85 @@ export const useVideoStore = defineStore('video', () => {
+    switch (autoRecordOption) {
+      case 'all':
+        startRecordingAllStreams()
+        openSnackbar({
```
This one as well.

ArturoManzoli commented:
@rafaellehmkuhl , 
- Rolled back the object name to the original and changed the propriety to the suggested one;
- Changed to momentary and persistent configs on the video.ts store and vue component;
- Fixed the snackbar on single line.

rafaellehmkuhl commented on src/stores/video.ts:
```diff
@@ -57,7 +59,14 @@ export const useVideoStore = defineStore('video', () => {
   const autoProcessVideos = useBlueOsStorage('cockpit-auto-process-videos', true)
   const lastRenamedStreamName = ref('')
-
+  const cockpitAutoRecordStreams = useBlueOsStorage<AutoRecordVideoStreams>(
+    'cockpit-auto-record-streams',
+    defaultAutoRecordingOptions
+  )
+  const currentautoRecordModes = ref(
+    cockpitAutoRecordStreams.value || { autoRecordMode: undefined, selectedStreams: [] }
+  )
+  const vehicleStore = useMainVehicleStore()
```
The names were not changed...

rafaellehmkuhl commented on src/components/RecordingReminder.vue:
```diff
@@ -0,0 +1,101 @@
+const bgColor = ref('bg-red-500')
+const borderRadius = ref(25)
+
+const initial_delay = 100
+const transition_to_rect_delay = 1500
+const fading_delay = 5000
+const close_delay = 5500
+const start_border_animation = 1000
+const border_animation_delay = 20
+const border_radius_start = 25
+const border_radius_end = 8
```
Javascript variables should always follow snakeCase.

I'm with the impression that the linter is completely broken, not just partially...

ArturoManzoli replied:
Yep, linter and also the TS checker

ArturoManzoli replied:
> Javascript variables should always follow snakeCase.
> 
> I'm with the impression that the linter is completely broken, not just partially...

Done. Case-converted

rafaellehmkuhl reviewed (changes requested):
The code is good now and the functionalities are all working.

There's one big problem thought, probably related to the separation between the persistent and momentary settings: changes to the settings are not being reactively saved in the ref variable that uses the settingsSyncer composable, and thus any changes made (and saved) by the user are not being sent to the vehicle during that session, but only in the next one, causing an automatic change on startup, and thus a conflict, both undesirable.

ArturoManzoli commented:
> The code is good now and the functionalities are all working.
> 
> There's one big problem thought, probably related to the separation between the persistent and momentary settings: changes to the settings are not being reactively saved in the ref variable that uses the settingsSyncer composable, and thus any changes made (and saved) by the user are not being sent to the vehicle during that session, but only in the next one, causing an automatic change on startup, and thus a conflict, both undesirable.

I think see what you mean.

In my understanding, copying the modified settings (1) object to the variable (2) that uses the blueOsStorage/settingsSyncer would be enough to write the new parameters on the vehicle.

````
(1) 
const handleCloseModal = (): void => {
  let currentOptions = { ...optionsSelectedByTheUser }

  **if (rememberChoice.value) {
    cockpitAutoRecordStreams.value = currentOptions
  }**

  videoStore.momentaryAutoRecordOptions = currentOptions
  visible.value = false
  emit('close')
}
````

`(2) const cockpitAutoRecordStreams = useBlueOsStorage<AutoRecordVideoStreams>('cockpit-auto-record-streams', defaultAutoRecordingOptions)`

Should it be done in a different way?
Ore is there something else interfering with the data saving?

rafaellehmkuhl commented:
> > The code is good now and the functionalities are all working.
> > There's one big problem thought, probably related to the separation between the persistent and momentary settings: changes to the settings are not being reactively saved in the ref variable that uses the settingsSyncer composable, and thus any changes made (and saved) by the user are not being sent to the vehicle during that session, but only in the next one, causing an automatic change on startup, and thus a conflict, both undesirable.
> 
> I see what you mean.
> 
> In my understanding, copying the modified settings object (1) to the variable (2) that uses the blueOsStorage/settingsSyncer would be enough to write the new parameters on the vehicle.
> 
> ```
> (1) 
> const handleCloseModal = (): void => {
>   let currentOptions = { ...optionsSelectedByTheUser }
> 
>   **if (rememberChoice.value) {
>     cockpitAutoRecordStreams.value = currentOptions
>   }**
> 
>   videoStore.momentaryAutoRecordOptions = currentOptions
>   visible.value = false
>   emit('close')
> }
> ```
> 
> `(2) const cockpitAutoRecordStreams = useBlueOsStorage<AutoRecordVideoStreams>('cockpit-auto-record-streams', defaultAutoRecordingOptions)`
> 
> Should it be done in a different way? Or is there something else interfering with the data saving?

It's probably a reactivity problem, since you're replacing the object instead of its options or instead of reassigning its values (take a look on similar implementations happen in the codebase, using `Object.assign`.

Also, I just noticed you actually have 3-4ish sources of truth for all those settings, and you also have two variables connected to the `useBlueOsStorage` composable. There's no need for it and it will more likely cause conflicts between those two states.

My suggestion: keep only the two variables you have in the store (`persistentAutoRecordOptions` and `momentaryAutoRecordOptions`), and use the momentary one directly in the dialog component. You don't need other variables in the component, you can directly connect the inputs on the momentary options. Using more than those two escalates the complexity of the feature quickly and without need, and any maintenance on it (like the one currently happening) gets complicated.

ArturoManzoli commented:
> > The code is good now and the functionalities are all working.
> > There's one big problem thought, probably related to the separation between the persistent and momentary settings: changes to the settings are not being reactively saved in the ref variable that uses the settingsSyncer composable, and thus any changes made (and saved) by the user are not being sent to the vehicle during that session, but only in the next one, causing an automatic change on startup, and thus a conflict, both undesirable.

Also, on a test version, removing the momentary options states and always saving the preferences on the videoStore's autoRecordOptions variable (that is linked to blueosStorage via useBluosStorage composable), I noticed the same problem. The settings are not being immediately stored to the vehicle. Only after a refresh

Do you have any thoughts on that?

rafaellehmkuhl commented:
> > > The code is good now and the functionalities are all working.
> > > There's one big problem thought, probably related to the separation between the persistent and momentary settings: changes to the settings are not being reactively saved in the ref variable that uses the settingsSyncer composable, and thus any changes made (and saved) by the user are not being sent to the vehicle during that session, but only in the next one, causing an automatic change on startup, and thus a conflict, both undesirable.
> 
> Also, on a test version, removing the momentary options states and always saving the preferences on the videoStore's autoRecordOptions variable (that is linked to blueosStorage via useBluosStorage composable), I noticed the same problem. The settings are not being immediately stored to the vehicle. Only after a refresh
> 
> Do you have any thoughts on that?

Yup. It's on the previous comment.

ArturoManzoli commented:
Last patch:
- Implemented suggestions about multiple sources of truth and keeping a single variable on sync with BlueOS;
- Lint fix.

ArturoManzoli committed:
Video: Add option to auto start stream recordings on vehicle arm

Signed-off-by: Arturo Manzoli <arturomanzoli@gmail.com>

Video: Auto-record-dialog: Lint fix

Signed-off-by: Arturo Manzoli <arturomanzoli@gmail.com>

ArturoManzoli commented:
> @ArturoManzoli could you squash the commits and also rebase over master (as there are conflicts with it) before I test? Just to make sure I'm running the final version.

@rafaellehmkuhl Done!

rafaellehmkuhl reviewed (changes requested):
Arturo, I'm having a pretty consistent behavior with this PR causing my vehicle to disarm when I open Cockpit.

The test is: arm the vehicle, restart Cockpit, it will disarm during Cockpit's boot.

The behavior as well as my auto-record configs can be seen in the video below:

https://github.com/user-attachments/assets/8dc72b38-74d8-466c-8b96-8f716ab2bd4b

It happens after I arm the vehicle within this branch for the first time.

I cannot reproduce the same behavior on master.

ArturoManzoli commented:
> Arturo, I'm having a pretty consistent behavior with this PR causing my vehicle to disarm when I open Cockpit.
> 
> The test is: arm the vehicle, restart Cockpit, it will disarm during Cockpit's boot.
> 
> The behavior as well as my auto-record configs can be seen in the video below:
> 
> It happens after I arm the vehicle within this branch for the first time.
> 
> I cannot reproduce the same behavior on master.

Hmm, that's odd. I'll replicate here and investigate

ArturoManzoli commented:
> Arturo, I'm having a pretty consistent behavior with this PR causing my vehicle to disarm when I open Cockpit.
> 
> The test is: arm the vehicle, restart Cockpit, it will disarm during Cockpit's boot.
> 
> The behavior as well as my auto-record configs can be seen in the video below:
> 
> It happens after I arm the vehicle within this branch for the first time.
> 
> I cannot reproduce the same behavior on master.

I was able to see the disarming once, on the first run, but never again.
I tried switching branches, reloading the server, resetting settings, deleting cockpit profiles on the bag of holding...

Do you have any idea what else can be done to reproduce that?

ArturoManzoli commented:
@rafaellehmkuhl I wasn't able to reproduce the disarming bug anymore here. Were you force arming?

rafaellehmkuhl commented:
> @rafaellehmkuhl I wasn't able to reproduce the disarming bug anymore here. Were you force arming?

This bug appears to have disappeared! Nice!

Right now there are only two things that are happening that need to be resolved:

1. If you have more than one video stream, one of them will not be recorded, even with the correct option selected.
2. Not captured in this video, but if the video stream takes sometime to load, the user will receive a "Media stream not loaded" warning dialog. This happened to me with the Radcam apparently. It does not happen all the time, but when it happens, it appears like the auto-record does not work.

https://github.com/user-attachments/assets/be0f9efc-d7cf-4e96-8fb7-f5e932497f15

rafaellehmkuhl reviewed (changes requested):
Right now there are only two things that are happening that need to be resolved:

1. If you have more than one video stream, one of them will not be processed.
2. Not captured in this video, but if the video stream takes sometime to load, the user will receive a "Media stream not loaded" warning dialog. This happened to me with the Radcam apparently. It does not happen all the time, but when it happens, it appears like the auto-record does not work.

https://github.com/user-attachments/assets/be0f9efc-d7cf-4e96-8fb7-f5e932497f15
